File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        ## create a dummy node 
        dummy = ListNode()
        tail = dummy
        
        ## both linked lists should be not null
        while(list1 and list2):
            ## compare both the values in the lists and place the small one into the result which is tail
            if(list1.val < list2.val):
                tail.next = list1
                ## increament the pointer
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
                
            tail = tail.next
            
        ## if one of the list is non null
        if(list1):
            tail.next = list1
        elif(list2):
            tail.next = list2
            
        return dummy.next    
        

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 
 
 // Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
};


 * ITERATIVE SOLUTION
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode();
        ListNode* tail = dummy;
        
        while(list1 && list2){
            if(list1->val < list2->val){
                tail->next = list1;
                list1 = list1->next;
            }
            else{
                tail->next = list2;
                list2 = list2->next;
            }
            
            tail = tail->next;
        }
        
        if(list1){
            tail->next = list1;
        }
        else if(list2){
            tail->next = list2;
        }
        
        // because the first node is dummy and the next chain of it         // is our answer
        return dummy->next;
    }
};